### ISSUE TITLE
"no-undef" rule and "browser" environment acting strange with regard to identifier case

### ISSUE THREAD

**Issue by nmschulte** [eslint/eslint#3735]
**What version are you using?**
```bash
$ eslint -v
v1.3.1
```

**What did you do?**
```bash
$ echo "var Document = {};" | eslint -f json --stdin --no-eslintrc --env "browser" --rule '{"no-undef": 2}'; echo $?
```
**What happened?**
```json
[{"filePath":"<text>","messages":[{"ruleId":"no-undef","severity":2,"message":"\"Document\" is read only.","line":1,"column":5,"nodeType":"Identifier","source":"var Document = {};"}],"errorCount":1,"warningCount":0}]
1
```

**What did you expect to happen?**
```json
[{"filePath":"<text>","messages":[],"errorCount":0,"warningCount":0}]
0
```

--

Oddly enough, if one adds the "commonjs" environment to the command invocation, the expected output is shown.  I have no idea why adding the "commonjs" environment makes this go away from the CLI, but my actual use-case is invoking ESLint with an hierarchical inherited `.eslintrc` configuration via webpack, and the file where this issue came up is configured to lint with both the "browser" and the "commonjs" environments, and I still receive the error there.

I think it's safe to ignore this CLI w/o `.eslintrc` vs webpack w/ `.eslintrc` oddity and just fix the case-sensitivity issue with "browser" environment/globals/the no-undef rule.

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by nmschulte** [comment]
Okay, I think now I am entirely wrong about this being an issue with `no-undef` and "browser".  I realize now that `Document` is indeed a browser environment "global" that should not be redefined (nor shadowed, if that's possible).

I think instead, this is an issue with the "commonjs" environment, and perhaps an issue with how environments are configured/applied when using the hierarchical inherited `.eslintrc` configuration.

I think that the expected behavior when adding the "commonjs" environment from the CLI is correct, but via webpack is not.

I'm not sure how to proceed; please advise!

**Comment by ilyavolodin** [comment]
I can reproduce this in online demo with `browser` environment vs. `browser + node` environment... This is pretty strange.

**Comment by mysticatea** [comment]
I will try to fix this.

**Comment by mysticatea** [comment]
Hmm, I remembered #3390.
In environments which have an extra scope (commonjs, node, and modules), this behavior might be an intended behavior.

**Comment by nmschulte** [comment]
@mysticatea I think the error should go away when we add the "commonjs" environment, yes.  I apologize, but I think my second posting has the real issue.

I think the real issue is that the `.eslintrc` setup (per-dir w/ extra / changing envs), doesn't pick up on that properly for some reason.  I can try to create a repo that reproduces this, but it will be a minute before I can do that.

**Comment by mysticatea** [comment]
`Window` and `Document` are an built in global variable (those type is `function`).
`window` is an object which has `Window.prototype` as its `__proto__` value.

Though I'm not sure why the current `no-undef` behavior reports redeclarations....

**Comment by ilyavolodin** [comment]
Pretty sure that's due to the historical reasons. `no-undef` was one of the first rules. We didn't have `no-redeclare` at that point yet. But maybe we should revisit this now?

**Comment by mysticatea** [comment]
@ilyavolodin Thank you for the explanation.
I guess it needs difficult arrangement, so I would like to open new issue before 2.0.

**Comment by ilyavolodin** [comment]
@mysticatea :+1: 
Back to the original issue. So now with @mysticatea explanation I understand why `browser` + `commonjs` environments are not complaining about redeclaration of `Document`, but I still don't think that's correct. Both environments suggest that you will run code in both places (not necessarily at the same time), so when you are running it in browser, you should not be redeclaring `Document`, even if under commonjs that might be a reasonable thing to do. I have no idea how we implement this, since commonjs creates an extra top level scope.

**Comment by nmschulte** [comment]
Creating a `var Document = {}` in the browser + CommonJS environments should cause a `no-shadow` issue, not a "overwriting a read-only global" issue.  I think that's perfectly acceptable; if you shadow a global, that's your problem.

**Comment by mysticatea** [comment]
Actually `no-redeclare` and `no-shadow` are handling the extra scope as global.

```js
/* eslint-env commonjs */
/* eslint no-redeclare:[2,{builtinGlobals:true}], no-shadow:[2,{builtinGlobals:true}] */
var Object = 0;
```

```
test.js
  3:5  error  Object is already defined  no-redeclare

✖ 1 problem (1 error, 0 warnings)
```

But also I understand the request in #3390...
Built-in globals includes many general names, especially in browsers.
So if somebody are 100% sure extra scope is made, they expect to not handle it as redeclaration.

(Ah, this is "difficult arrangement" I said before)

**Comment by nzakas** [comment]
I'm not sure there's a bug here. Can someone clarify?

**Comment by ilyavolodin** [comment]
So I think there's a bunch of unrelated issues (like `no-undef` reporting errors that it shouldn't have any business checking in the first place), but the main issue is that it's reported that behavior of environments inline is different from behavior inside `.eslintrc`. I just tested out the following `.eslintrc` file:
```json
{
	"env": {
        "browser": true,
        "commonjs": true
	},
	"rules": {
		"no-undef": 2
	}
}
```
and the following js file
```js
var Document = 32;
```
And I'm not getting any errors. So it looks like I can't reproduce the original issue. @nmschulte could you verify my test please?

**Comment by nmschulte** [comment]
First, apologies for the confusion; that's where I started from (confusion).

I think there are a few unrelated issues here, and the title of the issue is not one of them (nor is the first post, but it sets the stage).

@ilyavolodin I will try to isolate the environment issue, as I think that's the most important one here and the one I understand the most.  It sounds like the other issues require more knowledge of the inner workings of ESLint to really describe, and I'm not that familiar.

**Comment by mysticatea** [comment]
I am sorry that I confuse you. I misread many...
Hmm, I tried to reproduce that, but I have not reproduced yet.

**Comment by nmschulte** [comment]
This should be a test case for a couple of the issues here; first, the environments via inherited .eslintrc issue, and the no-undef overconcerned issue.

https://github.com/nmschulte/eslint-no_undef-environment-inherited_eslintrc_config

**Comment by mysticatea** [comment]
OK, thanks a lot @nmschulte. 
I found the minimum environment which reproduces it.

**~/.eslintrc**

```json
{
    "ecmaFeatures": {"globalReturn": false},
    "rules": {"no-undef": 2}
}
```

**~/src/.eslintrc**

```json
{
    "env": {"commonjs": true}
}
```

In this case, the `commonjs` environment should overwrite the `globalReturn` feature with `true`, but I got `false`.

**Comment by nmschulte** [comment]
So, I'm confused: aren't all of the `ecmaFeatures` for ES6+ features?  How does the `globalReturn` feature resolve this (and why does setting the `commonjs` environment flip this on)?  In reality, doesn't the CommonJS wrapper keep this from actually being a "global return"?  I'm not familiar with how ESLint works under the hood, but I imagine it literally stuffs the src in a function wrapper, e.g. `function() { /* INSERT_SRC */ }`; no?

Thanks for tracking this down.

**Comment by mysticatea** [comment]
Actually `ecmaFeatures` is espree's option, so it's parser switches. It includes es6, jsx, and some relaxing options. `globalReturn` is to allow `return` statements in the global. And yes, this option creates extra function scope for global variables.

`env` is composit option. It's converted to `ecmaFeatures` and definitions of implicit global variables.
https://github.com/eslint/eslint/blob/master/conf/environments.js#L29-L34
